lib/handler/connect.c

```diff
@@ -811,27 +820,28 @@ static void udp_on_read(h2o_socket_t *_sock, const char *err)
 
     /* read UDP packet, or return */
     ssize_t rret;
-    while ((rret = recv(h2o_socket_get_fd(self->sock), self->udp.ingress.buf + UDP_CHUNK_OVERHEAD,
-                        sizeof(self->udp.ingress.buf) - UDP_CHUNK_OVERHEAD, 0)) == -1 &&
-           errno == EINTR)
-        ;
+    int overhead = (self->src_req->datagram_format == H2O_DATAGRAM_FORMAT_DRAFT03) ? UDP_CHUNK_OVERHEAD_DRAFT03 : UDP_CHUNK_OVERHEAD_RFC;
+    while ((rret = recv(h2o_socket_get_fd(self->sock), self->udp.ingress.buf + overhead,
+                        sizeof(self->udp.ingress.buf) - overhead, 0)) == -1 && errno == EINTR);
     if (rret == -1)
         return;
     H2O_PROBE_REQUEST(CONNECT_UDP_READ, self->src_req, (size_t)rret);
 
     /* forward UDP datagram as is; note that it might be zero-sized */
     if (self->src_req->forward_datagram.read_ != NULL) {
-        h2o_iovec_t vec = h2o_iovec_init(self->udp.ingress.buf + UDP_CHUNK_OVERHEAD, rret);
+        h2o_iovec_t vec = h2o_iovec_init(self->udp.ingress.buf + overhead, rret);
         self->src_req->forward_datagram.read_(self->src_req, &vec, 1);
     } else {
         h2o_socket_read_stop(self->sock);
         h2o_timer_unlink(&self->timeout);
         size_t off = 0;
-        self->udp.ingress.buf[off++] = 0; /* chunk type = UDP_PACKET */
+        self->udp.ingress.buf[off++] = 0; /* chunk type = DATAGRAM (draft03) or capsule type = DATAGRAM (rfc) */
         off = quicly_encodev(self->udp.ingress.buf + off, (uint64_t)rret) - self->udp.ingress.buf;
```

Comment:
The capsule length that is being encoded here needs to include the context_id field because the context_id fields is inside the capsule (in front of the actual packet data).  Please see doohara/h2o#1.

Rationale: [RFC 9297 Section 3.5](https://datatracker.ietf.org/doc/html/rfc9297#section-3.5) specifies the format of the Datagram Capsule with a protocol-specific "HTTP Datagram Payload" inside.  [RFC 9298 Section 5](https://datatracker.ietf.org/doc/html/rfc9298#section-5) then specifies the format of the "HTTP Datagram Payload" in the case when HTTP is used for proxying UDP.  The "HTTP Datagram Payload" in that case consists of a "Context ID" followed by the "UDP Proxying Payload".  The length of the "HTTP Datagram Payload" is thus the length of the "UDP Proxying Payload" plus the length of the encoded "Context ID" field.